### PR TITLE
Add custom_dimensions using kwargs instead of args

### PR DESCRIPTION
**Pull request by xinyi-joffre**
This PR allows users to add customDimensions to their logs for both traces and exceptions. This approach was discussed in #834 before implementing.

As a part of this shift, it will break the changes introduced as part of #822. However, the new implementation allows for a smoother transition for those wanting to keep existing logging functionality with string interpolation and args. In addition, this PR adds support for exceptions which was not covered as part of the previous implementation.

Usage:

```python
import logging

from opencensus.ext.azure.log_exporter import AzureLogHandler

logger = logging.getLogger(__name__)
# TODO: set APPLICATIONINSIGHTS_CONNECTION_STRING as environment variable.
logger.addHandler(AzureLogHandler())

properties = {'custom_dimensions': {'key_1': 'value_1', 'key_2': 'value_2'}}

# Use properties in logging statements
logger.warning('action', extra=properties)

# Use properties in exception logs
try:
    result = 1 / 0  # generate a ZeroDivisionError
except Exception:
    logger.exception('Captured an exception.', extra=properties)

```

Once this data is uploaded to App Insights, you can search in the traces and exceptions table by keys included in the customDimensions column like this:

```
traces
| where custom_dimensions.key_1 == "value_1"
| limit 50
| order by timestamp desc
```

```
exceptions
| where custom_dimensions.key_1 == "value_1"
| limit 50
| order by timestamp desc
```

Finally, this PR also contains contextual logging examples, for those who need to add context to a large number of logs without needing to add the `extra=` keyword to each log statement.

### CODE REVIEW

**Comment by xinyi-joffre** [comment]
The CircleCI tests seem to be failing on `ModuleNotFoundError: No module named '_sqlite3'`. I don't think I changed anything that would break the existing modules. Could this be an existing issue with the build? @lzchen

File: contrib/opencensus-ext-azure/examples/logs/properties.py [hunk 1]
```diff
@@ -21,4 +21,14 @@
 # and place it in the APPLICATIONINSIGHTS_CONNECTION_STRING
 # environment variable.
 logger.addHandler(AzureLogHandler())
-logger.warning('action', {'key-1': 'value-1', 'key-2': 'value2'})
+
+properties = {'customDimensions': {'key-1': 'value-1', 'key-2': 'value2'}}
```
Comment (lzchen) [inline contrib/opencensus-ext-azure/examples/logs/properties.py]: I would follow the [PEP8](https://www.python.org/dev/peps/pep-0008/#function-and-variable-names) standard of naming variables using lower case and underscores.

Comment (xinyi-joffre) [reply contrib/opencensus-ext-azure/examples/logs/properties.py]: Thanks! I will update!

**Comment by xinyi-joffre** [comment]
@lzchen, do you want to me also keep the deprecated args implementation as well? I can add it, and print out a warning message about new functionality

File: contrib/opencensus-ext-azure/examples/logs/contextual_logging.py [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+            return {}
+
+    def process(self, msg, kwargs):
+
+        if self.custom_dimensions:
+            if 'extra' not in kwargs:
+                kwargs['extra'] = {'custom_dimensions': self.custom_dimensions}
+
+            elif 'extra' in kwargs and 'custom_dimensions' in kwargs['extra']:
+                kwargs['extra'] = {
```
Comment (lzchen) [inline contrib/opencensus-ext-azure/examples/logs/contextual_logging.py]: Would this overwrite other entries in the `extra` dict?

File: contrib/opencensus-ext-azure/examples/logs/contextual_logging.py [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+                    'custom_dimensions': {
+                        **self.custom_dimensions,
+                        **kwargs['extra']['custom_dimensions']
+                    }
+                }
+
```
Comment (lzchen) [inline contrib/opencensus-ext-azure/examples/logs/contextual_logging.py]: I think there's the missing case of if `extra` is in `kwargs` but `custom_dimensions` is not.

File: contrib/opencensus-ext-azure/examples/logs/contextual_logging.py (hunk 2 above)
Comment (xinyi-joffre) [reply contrib/opencensus-ext-azure/examples/logs/contextual_logging.py]: Good catch! You are right!

File: contrib/opencensus-ext-azure/examples/logs/contextual_logging.py [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+        if hasattr(record, 'custom_dimensions'):
+            record.custom_dimensions.update(self.custom_dimensions)
+        else:
+            setattr(record, 'custom_dimensions', self.custom_dimensions)
+
+        return True
+
+
+f = CustomDimensionsFilter(custom_dimensions={
```
Comment (lzchen) [inline contrib/opencensus-ext-azure/examples/logs/contextual_logging.py]: These are great examples!

**Comment by lzchen** [comment]
> @lzchen, do you want to me also keep the deprecated args implementation as well? I can add it, and print out a warning message about new functionality

I think we want to introduce this as the proper method to be adding custom dimensions, so it's okay to make this a breaking change.

**Comment by lzchen** [comment]
@xinyi-joffre  
Looks good and thanks for contributing! I will be taking a look at the build errors.

File: contrib/opencensus-ext-azure/examples/logs/contextual_logging.py (hunk 2 above)
Comment (xinyi-joffre) [reply contrib/opencensus-ext-azure/examples/logs/contextual_logging.py]: Another questions is that this is currently a sample of how someone can add contextual logging easily to their code, however, they would still need to implement these classes and support them (even though they are just related to updating a properties field for log export. Is it worthwhile to consider supporting these two classes as part of the package?

Comment (lzchen) [reply contrib/opencensus-ext-azure/examples/logs/contextual_logging.py]: The `Azure` package supports solely on features that relate to sending telemetry to Azure Monitor. How the user wants to do this is up to them. We might make the `properties` feature more robust in the future (by adding the classes as you've mentioned) but I think for now it is fine to just include them in the example file.

**Comment by lzchen** [comment]
@xinyi-joffre
Actually now that I think about it, introducing these classes in the examples might cause confusion to users (they might assume these exist as features). I think it will be better to remove this example file.

File: contrib/opencensus-ext-azure/examples/logs/contextual_logging.py (hunk 2 above)
Comment (xinyi-joffre) [reply contrib/opencensus-ext-azure/examples/logs/contextual_logging.py]: Makes sense!

File: contrib/opencensus-ext-azure/examples/logs/contextual_logging.py (hunk 4 above)
Comment (xinyi-joffre) [reply contrib/opencensus-ext-azure/examples/logs/contextual_logging.py]: Thank you 👍

**Comment by xinyi-joffre** [comment]
I also realized that the opencensus also uses 'attributes' in tracing for contextual information. Do you like the custom_dimensions name (since it matches the Application Insights field name)? Or should I change it to something simpler like attributes, properties, etc.?

**Comment by lzchen** [comment]
@xinyi-joffre 
Since this feature is specific to `Azure`, I would prefer to use `custom_dimensions`. `attributes` is taken from `span_attributes` which is a concept more related to tracing, so it is fine for logging to use a different name (metrics uses `tags` to represent custom dimensions). Also, `attributes` is a common enough word that users might be using it for other things.

**Comment by lzchen** [comment]
@xinyi-joffre 
Can you rebase? I believe [#842] has fixed the build errors.

**Comment by xinyi-joffre** [comment]
@lzchen this PR is ready to go, and passed the checks! Please let me know if you want me to fix anything else :). I can also help update the docs.microsoft.com entry for this once this is released.

**Comment by lzchen** [comment]
@xinyi-joffre 
Thanks for contributing. I can handle updating the docs, so no worries :)

**Pull request merged**